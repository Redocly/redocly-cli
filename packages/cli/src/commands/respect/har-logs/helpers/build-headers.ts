/**
 * Support the three possible header formats we'd get from a request or
 * response:
 *
 * - A flat array with both names and values: [name, value, name, value, ...]
 * - An object with array values: { name: [value, value] }
 * - An object with string values: { name: value }
 */
export function buildHeaders(headers: any): any[] {
  const list = [];
  if (Array.isArray(headers)) {
    for (let i = 0; i < headers.length; i += 2) {
      list.push({
        name: headers[i],
        value: headers[i + 1],
      });
    }
  } else if (headers instanceof Map || headers.entries) {
    // Handle both Map and Headers objects (which have entries())
    for (const [name, value] of headers.entries()) {
      list.push({ name, value });
    }
  } else if (typeof headers === 'object') {
    for (const [name, value] of Object.entries(headers)) {
      list.push({ name, value: value as string });
    }
  }
  return list;
}
