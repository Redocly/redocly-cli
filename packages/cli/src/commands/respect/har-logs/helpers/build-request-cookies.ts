import * as cookie from 'cookie';

export function buildRequestCookies(headers: any): any[] {
  const cookies: any[] = [];
  for (const [header, headerValue] of Object.entries(headers)) {
    if (header.toLowerCase() === 'cookie') {
      // Handle both array and single string cases
      const cookieValues = Array.isArray(headerValue) ? headerValue : [headerValue];
      for (const value of cookieValues) {
        const parsed = cookie.parse(value);
        for (const [name, value] of Object.entries(parsed)) {
          cookies.push({ name, value });
        }
      }
    }
  }
  return cookies;
}
