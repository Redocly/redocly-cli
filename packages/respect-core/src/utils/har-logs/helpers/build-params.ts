import * as querystring from 'querystring';

export function buildParams(paramString: string): any[] {
  const params = [];
  const parsed = querystring.parse(paramString);
  for (const name in parsed) {
    const value = parsed[name];
    if (Array.isArray(value)) {
      value.forEach((item) => {
        params.push({ name, value: item });
      });
    } else {
      params.push({ name, value });
    }
  }
  return params;
}
