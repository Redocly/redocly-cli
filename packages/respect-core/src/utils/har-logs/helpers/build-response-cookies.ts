import * as setCookie from 'set-cookie-parser';

export function buildResponseCookies(headers: any): any[] {
  const cookies: any[] = [];
  // Handle both plain objects and Headers instance
  const setCookies =
    headers instanceof Headers
      ? headers.getSetCookie() // For Headers object
      : headers['set-cookie']; // For plain objects

  if (setCookies) {
    for (const headerValue of setCookies) {
      let parsed;
      try {
        parsed = setCookie.parse(headerValue);
      } catch {
        continue;
      }
      for (const cookie of parsed) {
        const { name, value, path, domain, expires, httpOnly, secure } = cookie;
        const harCookie: any = {
          name,
          value,
          httpOnly: httpOnly || false,
          secure: secure || false,
        };
        if (path) {
          harCookie.path = path as string | undefined;
        }
        if (domain) {
          harCookie.domain = domain as string | undefined;
        }
        if (expires) {
          harCookie.expires = expires.toISOString();
        }
        cookies.push(harCookie);
      }
    }
  }
  return cookies;
}
