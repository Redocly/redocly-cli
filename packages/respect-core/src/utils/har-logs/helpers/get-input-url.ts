export function getInputUrl(input: any): URL {
  const url = typeof input === 'string' ? input : input.url;
  return new URL(url);
}
