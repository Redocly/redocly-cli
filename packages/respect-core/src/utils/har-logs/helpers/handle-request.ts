import type { Dispatcher } from 'undici';
import type { Har, Entry } from 'har-format';
import type { IncomingHttpHeaders } from 'http';

interface HarEntry extends Entry {
  _timestamps?: {
    start: number;
    connect: number;
    firstByte: number;
  };
}

export function handleRequest({
  input,
  handler,
  harLog,
  isUndici = false,
}: {
  input: Request | Dispatcher.DispatchOptions;
  handler: Dispatcher.DispatchHandlers | NodeJS.EventEmitter;
  harLog?: Map<string, Har>;
  isUndici?: boolean;
}): Dispatcher.DispatchHandlers | NodeJS.EventEmitter {
  if (isUndici) {
    return handleUndiciRequest(
      input as Dispatcher.DispatchOptions,
      handler as Dispatcher.DispatchHandlers,
      harLog
    );
  } else {
    handleNodeRequest(input as Request, handler as NodeJS.EventEmitter, harLog);
    return handler as NodeJS.EventEmitter;
  }
}

function handleUndiciRequest(
  options: Dispatcher.DispatchOptions,
  handler: Dispatcher.DispatchHandlers,
  harLog?: Map<string, Har>
): Dispatcher.DispatchHandlers {
  const startTime = Date.now();
  const requestId = generateRequestId();

  const harEntry: HarEntry = {
    startedDateTime: new Date().toISOString(),
    time: -1,
    request: {
      method: options.method || 'GET',
      url: options.path || '',
      httpVersion: 'HTTP/1.1',
      cookies: [],
      headers: formatUndiciHeaders(options.headers as any),
      queryString: [],
      headersSize: -1,
      bodySize: -1,
    },
    response: {
      status: 0,
      statusText: '',
      httpVersion: 'HTTP/1.1',
      cookies: [],
      headers: [],
      content: {
        size: 0,
        mimeType: '',
        text: '',
      },
      redirectURL: '',
      headersSize: -1,
      bodySize: -1,
    },
    cache: {},
    timings: {
      blocked: -1,
      dns: -1,
      connect: -1,
      send: 0,
      wait: 0,
      receive: 0,
      ssl: -1,
    },
  };

  const wrappedHandler: Dispatcher.DispatchHandlers = {
    onConnect: handler.onConnect,
    onError: (error) => {
      harEntry.response.status = 0;
      harEntry.response.statusText = error.message;
      harEntry.time = Date.now() - startTime;
      if (harLog) {
        harLog.set(requestId, {
          log: {
            version: '1.2',
            creator: { name: 'api-test-framework', version: '1.0' },
            entries: [harEntry],
          },
        });
      }
      if (handler.onError) {
        handler.onError(error);
      }
    },
    onHeaders: (statusCode, headers, resume, opaque) => {
      harEntry.response.status = statusCode;
      harEntry.response.headers = formatUndiciHeaders(headers as any);
      if (handler.onHeaders) {
        handler.onHeaders(statusCode, headers, resume, opaque);
      }
      return true;
    },
    onData: handler.onData,
    onComplete: (trailers) => {
      harEntry.time = Date.now() - startTime;
      if (harLog) {
        harLog.set(requestId, {
          log: {
            version: '1.2',
            creator: { name: 'api-test-framework', version: '1.0' },
            entries: [harEntry],
          },
        });
      }
      if (handler.onComplete) {
        handler.onComplete(trailers);
      }
    },
  };

  return wrappedHandler;
}

function formatUndiciHeaders(
  headers: Record<string, string | string[]> | string[] | IncomingHttpHeaders | null | undefined
): Array<{ name: string; value: string }> {
  if (!headers) return [];

  if (Array.isArray(headers)) {
    return chunks(headers, 2).map(([name, value]) => ({
      name: String(name),
      value: String(value),
    }));
  }

  return [];
}

function chunks<T>(arr: T[], size: number): T[][] {
  return Array.from({ length: Math.ceil(arr.length / size) }, (_, i) =>
    arr.slice(i * size, i * size + size)
  );
}

function generateRequestId(): string {
  return Math.random().toString(36).substring(2, 15);
}

function handleNodeRequest(request: Request, handler: any, harLog?: Map<string, Har>): void {
  const requestId = generateRequestId();
  const harEntry: Entry = {
    startedDateTime: new Date().toISOString(),
    time: 0,
    request: {
      method: request.method || 'GET',
      url: request.url,
      httpVersion: 'HTTP/1.1',
      cookies: [],
      headers: formatUndiciHeaders(request.headers as any),
      queryString: [],
      headersSize: -1,
      bodySize: -1,
    },
    response: {
      status: 0,
      statusText: '',
      httpVersion: 'HTTP/1.1',
      cookies: [],
      headers: [],
      content: {
        size: 0,
        mimeType: '',
        text: '',
      },
      redirectURL: '',
      headersSize: -1,
      bodySize: -1,
    },
    cache: {},
    timings: {
      blocked: -1,
      dns: -1,
      connect: -1,
      send: -1,
      wait: -1,
      receive: -1,
      ssl: -1,
    },
  };
  // Add HAR tracking header
  if (request.headers instanceof Headers) {
    request.headers.set('x-har-request-id', requestId);
  } else if (typeof request.headers === 'object' && request.headers !== null) {
    (request.headers as Record<string, string>)['x-har-request-id'] = requestId;
  }

  // Register all required event handlers
  handler.on('response', (response: any) => {
    harEntry.response.status = response.statusCode;
    harEntry.response.statusText = response.statusMessage;
    harEntry.response.headers = formatUndiciHeaders(response.headers);
  });

  handler.on('data', (chunk: Buffer) => {
    // Handle response data chunks
    harEntry.response.content.text = (harEntry.response.content.text || '') + chunk.toString();
    harEntry.response.content.size += chunk.length;
  });

  handler.on('end', () => {
    if (harLog) {
      harLog.set(requestId, {
        log: {
          version: '1.2',
          creator: { name: 'api-test-framework', version: '1.0' },
          entries: [harEntry],
        },
      });
    }
  });

  handler.on('error', (error: Error) => {
    harEntry.response.status = 0;
    harEntry.response.statusText = error.message;
    if (harLog) {
      harLog.set(requestId, {
        log: {
          version: '1.2',
          creator: { name: 'api-test-framework', version: '1.0' },
          entries: [harEntry],
        },
      });
    }
  });
}
