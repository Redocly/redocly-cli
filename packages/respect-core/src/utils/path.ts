import { isBrowser } from '@redocly/openapi-core';

let pathLib: any;

if (isBrowser) {
  // Browser environment
  pathLib = await import('path-browserify');
} else {
  // Node.js environment
  pathLib = await import('node:path');
}

export default pathLib;
