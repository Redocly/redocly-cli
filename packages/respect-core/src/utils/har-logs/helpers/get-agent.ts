import * as http from 'http';
import * as https from 'https';
import { Dispatcher } from 'undici';
import { getInputUrl } from './get-input-url';
import { handleRequest } from './handle-request';

const HarHttpAgent = createAgentClass(http.Agent);
const HarHttpsAgent = createAgentClass(https.Agent);

let globalHttpAgent: any;
let globalHttpsAgent: any;

// Add new Undici dispatcher
class HarDispatcher extends Dispatcher {
  constructor(opts?: any) {
    super(opts);
  }

  dispatch(options: Dispatcher.DispatchOptions, handler: Dispatcher.DispatchHandlers): boolean {
    // Handle HAR logging here similar to handleRequest
    handleRequest({ input: options, handler, harLog: new Map(), isUndici: true });
    return super.dispatch(options, handler);
  }
}

export function getAgent(input: any, options: any): any {
  // Add Undici dispatcher support
  if (options.dispatcher) {
    if (options.dispatcher instanceof Dispatcher) {
      return new HarDispatcher(options.dispatcher);
    }
    return options.dispatcher;
  }

  if (options.agent) {
    if (typeof options.agent === 'function') {
      return function (...args: any[]) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const agent = options.agent.call(this, ...args);
        if (agent) {
          instrumentAgentInstance(agent);
          return agent;
        }
        return getGlobalAgent(input);
      };
    }
    instrumentAgentInstance(options.agent);
    return options.agent;
  }
  return getGlobalAgent(input);
}

function getGlobalAgent(input: any): any {
  const url = getInputUrl(input);
  if (url.protocol === 'http:') {
    if (!globalHttpAgent) {
      globalHttpAgent = new HarHttpAgent();
    }
    return globalHttpAgent;
  }
  if (!globalHttpsAgent) {
    globalHttpsAgent = new HarHttpsAgent();
  }
  return globalHttpsAgent;
}

/**
 * Instrument an existing Agent instance. This overrides the instance's
 * `addRequest` method. It should be fine to continue using for requests made
 * without `withHar` - if the request doesn't have our `x-har-request-id`
 * header, it won't do anything extra.
 */
function instrumentAgentInstance(agent: any): void {
  const { addRequest: originalAddRequest } = agent;
  if (!originalAddRequest.isHarEnabled) {
    agent.addRequest = function addRequest(request: any, ...args: []) {
      handleRequest(request, ...args);
      return originalAddRequest.call(this, request, ...args);
    };
    agent.addRequest.isHarEnabled = true;
  }
}

function createAgentClass(BaseAgent: any): any {
  class HarAgent extends BaseAgent {
    constructor(...args: any[]) {
      super(...args);
      (this.addRequest as any).isHarEnabled = true;
    }

    addRequest(request: any, ...args: []): void {
      handleRequest(request, ...args);
      return super.addRequest(request, ...args);
    }
  }

  return HarAgent;
}
