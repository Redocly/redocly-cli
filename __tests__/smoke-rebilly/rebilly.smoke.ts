import { join } from 'node:path';
import { getCommandOutput, getParams } from '../helpers';

test('rebilly test case', () => {
  const indexEntryPoint = join(process.cwd(), 'packages/cli/lib/index.js');
  const fixturesPath = join(__dirname, 'rebilly.arazzo.yaml');
  const args = getParams(indexEntryPoint, ['respect', fixturesPath]);

  const result = getCommandOutput(args);
  expect(result).toMatchSnapshot();
});
