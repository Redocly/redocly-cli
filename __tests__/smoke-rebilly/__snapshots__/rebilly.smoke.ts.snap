// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rebilly test case 1`] = `
"────────────────────────────────────────────────────────────────────────────────

  Running workflow rebilly.arazzo.yaml / crud

  ✓ POST https://api-sandbox.rebilly.com/organizations/redocly/customers - step create-member
    ✓ success criteria check - $statusCode == 201

  ✓ GET https://api-sandbox.rebilly.com/organizations/redocly/customers/{id} - step read-member
    ✓ success criteria check - $statusCode == 200
    ✓ success criteria check - $response.body#/firstName == 'First'
    ✓ success criteria check - $response.body#/lastName == 'Last'

  ✓ PUT https://api-sandbox.rebilly.com/organizations/redocly/customers/{id} - step update-member
    ✓ success criteria check - $statusCode == 200
    ✓ success criteria check - $response.body#/firstName == 'UpdatedName'
    ✓ success criteria check - $response.body#/lastName == 'Last'
    ✓ success criteria check - $response.body#/id == $steps.create-member.outputs...

  ✓ DELETE https://api-sandbox.rebilly.com/organizations/redocly/customers/{id} - step delete-member
    ✓ success criteria check - $statusCode == 204

  ✓ GET https://api-sandbox.rebilly.com/organizations/redocly/customers/{id} - step read-non-existing-member
    ✓ success criteria check - $statusCode == 404

────────────────────────────────────────────────────────────────────────────────

  Running workflow rebilly.arazzo.yaml / inherited

  ✓ POST /customers - step add-user
    ✓ status code check - $statusCode in [201, 401, 403, 409, 422]
    ✓ content-type check
    ✓ schema check


  Summary for rebilly.arazzo.yaml
  
  Workflows: 2 passed, 2 total
  Steps: 6 passed, 6 total
  Checks: 13 passed, 13 total
  Time: <test>ms


┌─────────────────────────────────────────────────────────────┬────────────┬─────────┬─────────┬──────────┐
│ Filename                                                    │ Workflows  │ Passed  │ Failed  │ Warnings │
├─────────────────────────────────────────────────────────────┼────────────┼─────────┼─────────┼──────────┤
│ ✓ rebilly.arazzo.yaml                                       │ 2          │ 2       │ -       │ -        │
└─────────────────────────────────────────────────────────────┴────────────┴─────────┴─────────┴──────────┘


"
`;
